test/distributed/_shard/checkpoint/test_checkpoint.py

```diff
@@ -260,5 +283,227 @@ def test_create_key_handles_collision(self):
+    def __init__(
+        self,
+        fail_conf
+    ):
+        super(TestStorageWriter, self).__init__(fail_conf)
+
+    def prepare(self) -> None:
+        self._fail_rank("fail_prepare")
+
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        self._fail_rank("fail_write_bytes_on_ranks")
+        return self._fail_rank_async("fail_write_bytes_on_ranks_async")
+
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
+        self._fail_rank("fail_write_tensors_on_ranks")
+        return self._fail_rank_async("fail_write_tensors_on_ranks_async")
+
+    def finish(self, metadata: Metadata) -> None:
+        self._fail_rank("fail_finish")
+
+    def prepare_storage(self, storage_writes: List[Union[TensorWriteRequest, BytesWriteRequest]]) -> None:
+        self._fail_rank("fail_prepare_storage")
+
+class TestStorageReader(TestStorageBase, StorageReader):
+    def __init__(
+        self,
+        metadata,
+        fail_conf
+    ):
+        super(TestStorageReader, self).__init__(fail_conf)
+        self.metadata = metadata
+
+    def read_bytes(self, requests: List[BytesReadRequest]) -> Future[None]:
+        self._fail_rank("fail_read_bytes")
+        bad_ranks = self._get_ranks("fail_deser_bytes")
+        for r in requests:
+            if bad_ranks is not None and self.rank in bad_ranks:
+                # this is not "guaranteed" to fail, but hard to beat
+                rand = random.Random(1237)
+                r.bytes.write(rand.randbytes(32))
+            else:
+                torch.save([1, 2, 3], r.bytes)
+
+        return self._fail_rank_async("fail_read_bytes_async")
+
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
+        self._fail_rank("fail_read_tensors")
+        return self._fail_rank_async("fail_read_tensors_async")
+
+    def read_metadata(self) -> Metadata:
+        self._fail_rank("fail_read_metadata")
+        return self.metadata
+
+class TestDistributedFailure(ShardedTensorTestBase):
+    def get_spec(self):
+        return ChunkShardingSpec(
+            dim=0,
+            placements=[
+                f"rank:{r}/cuda:{r}" for r in range(dist.get_world_size())
+            ]
+        )
+
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(2)
+    @requires_nccl()
+    def test_dummy_writer_works(self) -> None:
+        state_dict = {
+            'sharded': sharded_tensor.rand(self.get_spec(), 20, 20),
+            'replicated': torch.rand(10, 10),
+            'bytes': [1, 2, 3, 4]
+        }
+
+        save_state_dict(state_dict, TestStorageWriter({}))
+
+
+    def _test_dist_failure(self, callback, kwargs):
+        bad_ranks = list(kwargs.values())[0] if len(kwargs) > 0 else []
+        try:
+            callback()
+            if len(bad_ranks) > 0:
+                self.fail(msg="Expected CheckpointException to be raised")
+        except CheckpointException as e:
```

Comment:
nit: Can use `self.assertRaises` instead?

Reply:
I didn't know you could get the exception with `assertRaises`. TIL